pilot/pkg/config/kube/ingress/ingress.go

```diff
@@ -0,0 +1,266 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingress
+
+import (
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"istio.io/istio/pkg/config/mesh/meshwatcher"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/log"
+	netutil "istio.io/istio/pkg/util/net"
+	corev1 "k8s.io/api/core/v1"
+	knetworking "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func SupportedIngresses(
+	ingress krt.Collection[*knetworking.Ingress],
+	ingressClass krt.Collection[*knetworking.IngressClass],
+	meshConfig meshwatcher.WatcherCollection,
+	services krt.Collection[*corev1.Service],
+	nodes krt.Collection[*corev1.Node],
+	pods krt.Collection[*corev1.Pod],
+	opts krt.OptionsBuilder,
+) (krt.Collection[krt.ObjectWithStatus[*knetworking.Ingress, knetworking.IngressStatus]], krt.Collection[*knetworking.Ingress]) {
+	podsByNamespace := krt.NewNamespaceIndex(pods)
+	return krt.NewStatusCollection(
+		ingress,
+		func(ctx krt.HandlerContext, i *knetworking.Ingress) (*knetworking.IngressStatus, **knetworking.Ingress) {
+			var class *knetworking.IngressClass
+			if i.Spec.IngressClassName != nil {
+				c := krt.FetchOne(ctx, ingressClass, krt.FilterKey(*i.Spec.IngressClassName))
+				if c == nil {
+					return nil, nil
+				}
+
+				class = *c
+			}
+
+			mesh := krt.FetchOne(ctx, meshConfig.AsCollection())
+			if !shouldProcessIngressWithClass(mesh.MeshConfig, i, class) {
+				// TODO: this is currently necessary to reset status of Ingresses which were previously selected
+				// but it seems overly expensive since it implies that we will always update the status of all Ingresses
+				// which are not selected. Status collection should eventually be cleaned because main collection receives a
+				// delete event for the Ingress.
+				return &knetworking.IngressStatus{}, nil
```

Comment:
It should detect a status going from <non-nil> to nil. Tested in https://github.com/istio/istio/pull/55926.

It might not make it all the way through the status queue, though.. I think it does, but the test doesn't cover that

Reply:
Also, the tests in #55926 only test for presence in the `Status` collection but they don't check if the status if really applied in k8s.

The issue is here: [pkg/kube/krt/status_test.go](https://github.com/istio/istio/blob/master/pilot/pkg/config/kube/gateway/status.go#L72-L73). If the an item is unselected from the collection, then we enqueue a status update with the previous Status (`Latest()` returns the last state of the object before it was deleted), I think here we should enqueue un update on the `Latest` object but with `nil` status. The only concern is that if the object was effectively deleted (and not unselected) then we would be unnecesarily attempting to update it's status.

Reply:
Ah I get what you mean. Yeah I think we need to make the registerStatus handle the delete case like you said